### ISSUE TITLE
Channels stuck in state AWAITING_UNILATERAL / signature missing in closing transaction

### ISSUE THREAD

**Issue by zerofeerouting** [ElementsProject/lightning#5734]
### Issue and Steps to Reproduce

After closing down my node, I have force closed all channels where the peer was disconnected after waiting for two weeks for them to come back online to cooperatively close.

Unfortunately almost all these force-closes (121 to be exact) are stuck in state `AWAITING_UNILATERAL` - it looks like the reason is the same for all of them: **The force-close transaction has no signature for the utxo input**.

# Example

## Close command

I initiated a close with a timeout of 2 weeks at `2022-11-01T11:45:02.232Z`

```
lightning-cli close -k id=28549x2160x1 unilateraltimeout=1209600
```

When after 15 days nothing happened, I issued a second close command - this time with a timeout of 1 second

```
lightning-cli close -k id=28549x2160x1 unilateraltimeout=1
```

Which resulted in the following force close:

## Force Close

```
# peer is offline, will negotiate once they reconnect (1 seconds before unilateral close).
# Timed out, forcing close.
{
   "tx": "0200000001dee7a0b8bd7f371af425b0484bf18addc60a0e2eb8e86822a509500663e83a5a0100000000bf9a1d80023322030000000000220020c60f059f123abe32e76dafeb167fd4f95ed350a8de578f500a1afed5342f0f14371b0c00000000001600147a64d41a7d297e62c1b9432c76f27668caa6edff4837b320",
   "txid": "cef138433dc464b617b00c8c1d9f0914d2a43657a6e4f701f81e50f787a06c3c",
   "type": "unilateral"
}
```
## Decoded FC transaction

```
{
  "txid": "cef138433dc464b617b00c8c1d9f0914d2a43657a6e4f701f81e50f787a06c3c",
  "hash": "cef138433dc464b617b00c8c1d9f0914d2a43657a6e4f701f81e50f787a06c3c",
  "version": 2,
  "size": 125,
  "vsize": 125,
  "weight": 500,
  "locktime": 548616008,
  "vin": [
    {
      "txid": "5a3ae863065009a52268e8b82e0e0ac6dd8af14b48b025f41a377fbdb8a0e7de",
      "vout": 1,
      "scriptSig": {
        "asm": "",
        "hex": ""
      },
      "sequence": 2149423807
    }
  ],
  "vout": [
    {
      "value": 0.00205363,
      "n": 0,
      "scriptPubKey": {
        "asm": "0 c60f059f123abe32e76dafeb167fd4f95ed350a8de578f500a1afed5342f0f14",
        "desc": "addr(bc1qcc8st8cj82lr9emd4l43vl75l90dx59gmetc75q2rtld2dp0pu2q7dxfu9)#fz039rth",
        "hex": "0020c60f059f123abe32e76dafeb167fd4f95ed350a8de578f500a1afed5342f0f14",
        "address": "bc1qcc8st8cj82lr9emd4l43vl75l90dx59gmetc75q2rtld2dp0pu2q7dxfu9",
        "type": "witness_v0_scripthash"
      }
    },
    {
      "value": 0.00793399,
      "n": 1,
      "scriptPubKey": {
        "asm": "0 7a64d41a7d297e62c1b9432c76f27668caa6edff",
        "desc": "addr(bc1q0fjdgxna99lx9sdegvk8dunkdr92dm0lzlup97)#azx60656",
        "hex": "00147a64d41a7d297e62c1b9432c76f27668caa6edff",
        "address": "bc1q0fjdgxna99lx9sdegvk8dunkdr92dm0lzlup97",
        "type": "witness_v0_keyhash"
      }
    }
  ]
}

```

## Channel `728549x2160x1`

https://mempool.space/lightning/channel/801048097046134785

```
        {
               "state": "AWAITING_UNILATERAL",
               "scratch_txid": "cef138433dc464b617b00c8c1d9f0914d2a43657a6e4f701f81e50f787a06c3c",
               "last_tx_fee": "1307000msat",
               "last_tx_fee_msat": "1307000msat",
               "feerate": {
                  "perkw": 1805,
                  "perkb": 7220
               },
               "short_channel_id": "728549x2160x1",
               "direction": 0,
               "channel_id": "dee7a0b8bd7f371af425b0484bf18addc60a0e2eb8e86822a509500663e83a5b",
               "funding_txid": "5a3ae863065009a52268e8b82e0e0ac6dd8af14b48b025f41a377fbdb8a0e7de",
               "funding_outnum": 1,
               "close_to_addr": "bc1q9vqg37d7e397yvqh9u8tqhvhvaxqp4cqytk5px",
               "close_to": "00142b0088f9becc4be230172f0eb05d97674c00d700",
               "private": false,
               "opener": "remote",
               "closer": "local",
               "features": [
                  "option_static_remotekey"
               ],
               "funding_allocation_msat": {
                  "03b64a8542852999fb2df9591162804313692dc1406db1e46619cbc7bac9a22195": 1000069000,
                  "038fe1bd966b5cb0545963490c631eaa1924e2c4c0ea4e7dcb5d4582a1e7f2f1a5": 0
               },
               "funding_msat": {
                  "03b64a8542852999fb2df9591162804313692dc1406db1e46619cbc7bac9a22195": "1000069000msat",
                  "038fe1bd966b5cb0545963490c631eaa1924e2c4c0ea4e7dcb5d4582a1e7f2f1a5": "0msat"
               },
               "funding": {
                  "local_msat": "0msat",
                  "remote_msat": "1000069000msat",
                  "pushed_msat": "0msat"
               },
               "msatoshi_to_us": 205363599,
               "to_us_msat": "205363599msat",
/* ... */
               "htlc_minimum_msat": 0,
               "minimum_htlc_in_msat": "0msat",
               "minimum_htlc_out_msat": "1msat",
               "maximum_htlc_out_msat": "1000msat",
               "their_to_self_delay": 144,
               "our_to_self_delay": 144,
               "max_accepted_htlcs": 30,
               "state_changes": [
                  {
                     "timestamp": "2022-03-22T19:19:48.749Z",
                     "old_state": "CHANNELD_AWAITING_LOCKIN",
                     "new_state": "CHANNELD_NORMAL",
                     "cause": "remote",
                     "message": "Lockin complete"
                  },
                  {
                     "timestamp": "2022-11-01T11:45:02.232Z",
                     "old_state": "CHANNELD_NORMAL",
                     "new_state": "CHANNELD_SHUTTING_DOWN",
                     "cause": "user",
                     "message": "User or plugin invoked close command"
                  },
                  {
                     "timestamp": "2022-11-16T13:14:15.678Z",
                     "old_state": "CHANNELD_SHUTTING_DOWN",
                     "new_state": "AWAITING_UNILATERAL",
                     "cause": "user",
                     "message": "Forcibly closed by `close` command timeout"
                  }
               ],
               "status": [
                  "AWAITING_UNILATERAL:Will attempt reconnect in 300 seconds"
               ],
/* ... */
               "htlcs": []
            }
```

## Manually sending the tx

Results in this error

```
error code: -26
error message:
non-mandatory-script-verify-flag (Witness program was passed an empty witness)
```

### `getinfo` output

`"version": "v0.11.2`

**Comment by rustyrussell** [comment]
Wow, this is cool!

These are pretty well tested code paths, but obviously we're not populating the witness at all.  Looking carefully, the only thing I can think of is that libwally is returning an error when we try to add the witness.  I've added some more error checking, so at least it should tell us why it's upset:

```diff
diff --git a/bitcoin/tx.c b/bitcoin/tx.c
index 29d0ed8ab..cbc27357c 100644
--- a/bitcoin/tx.c
+++ b/bitcoin/tx.c
@@ -318,33 +318,45 @@ void bitcoin_tx_output_get_amount_sat(const struct bitcoin_tx *tx, int outnum,
 	*amount = amount_asset_to_sat(&asset_amt);
 }
 
-void bitcoin_tx_input_set_witness(struct bitcoin_tx *tx, int innum,
-				  u8 **witness)
+const char *bitcoin_tx_input_set_witness(struct bitcoin_tx *tx, int innum,
+					 u8 **witness)
 {
 	struct wally_tx_witness_stack *stack = NULL;
 	size_t stack_size = tal_count(witness);
+	int ret;
 
 	tal_wally_start();
 	/* Free any lingering witness */
 	if (witness) {
-		wally_tx_witness_stack_init_alloc(stack_size, &stack);
-		for (size_t i = 0; i < stack_size; i++)
-			wally_tx_witness_stack_add(stack, witness[i],
-						   tal_bytelen(witness[i]));
+		ret = wally_tx_witness_stack_init_alloc(stack_size, &stack);
+		if (ret != WALLY_OK)
+			return tal_fmt(tx, "wally_tx_witness_stack_init_alloc returned %i", ret);
+		for (size_t i = 0; i < stack_size; i++) {
+			ret = wally_tx_witness_stack_add(stack, witness[i],
+							 tal_bytelen(witness[i]));
+			if (ret != WALLY_OK)
+				return tal_fmt(tx, "wally_tx_witness_stack_add %zi %s returned %i",
+				      i, tal_hex(tmpctx, witness[i]), ret);
+		}
 	}
 	tal_wally_end(tmpctx);
 
 	tal_wally_start();
-	wally_tx_set_input_witness(tx->wtx, innum, stack);
+	ret = wally_tx_set_input_witness(tx->wtx, innum, stack);
+	if (ret != WALLY_OK)
+		return tal_fmt(tx, "wally_tx_set_input_witness %i returned %i", innum, ret);
 	tal_wally_end(tx->wtx);
 
 	/* Also add to the psbt */
 	tal_wally_start();
-	wally_psbt_input_set_final_witness(&tx->psbt->inputs[innum], stack);
+	ret = wally_psbt_input_set_final_witness(&tx->psbt->inputs[innum], stack);
+	if (ret != WALLY_OK)
+		return tal_fmt(tx, "wally_psbt_input_set_final_witness %i returned %i", innum, ret);
 	tal_wally_end(tx->psbt);
 
 	if (taken(witness))
 		tal_free(witness);
+	return NULL;
 }
 
 void bitcoin_tx_input_set_script(struct bitcoin_tx *tx, int innum, u8 *script)
diff --git a/bitcoin/tx.h b/bitcoin/tx.h
index cb0903ccf..05bc32ddc 100644
--- a/bitcoin/tx.h
+++ b/bitcoin/tx.h
@@ -174,9 +174,11 @@ struct amount_asset bitcoin_tx_output_get_amount(const struct bitcoin_tx *tx,
  *
  * Given that we generate the witness after constructing the transaction
  * itself, we need a way to attach a witness to an existing input.
+ *
+ * Returns NULL on success, otherwise an error message allocated off tx.
  */
-void bitcoin_tx_input_set_witness(struct bitcoin_tx *tx, int innum,
-				  u8 **witness TAKES);
+const char *bitcoin_tx_input_set_witness(struct bitcoin_tx *tx, int innum,
+					 u8 **witness TAKES);
 
 /**
  * Set the input script on the given input.
diff --git a/lightningd/peer_control.c b/lightningd/peer_control.c
index bfd2e1e3e..bff93811c 100644
--- a/lightningd/peer_control.c
+++ b/lightningd/peer_control.c
@@ -186,6 +186,7 @@ static void sign_last_tx(struct channel *channel,
 	struct lightningd *ld = channel->peer->ld;
 	struct bitcoin_signature sig;
 	u8 *msg, **witness;
+	const char *err;
 
 	u64 commit_index = channel->next_index[LOCAL] - 1;
 
@@ -211,12 +212,17 @@ static void sign_last_tx(struct channel *channel,
 				 &sig, &channel->channel_info.remote_fundingkey,
 				 &channel->local_funding_pubkey);
 
-	bitcoin_tx_input_set_witness(last_tx, 0, take(witness));
+	err = bitcoin_tx_input_set_witness(last_tx, 0, take(witness));
+	if (err)
+		fatal("%s: %s", __func__, err);
 }
 
 static void remove_sig(struct bitcoin_tx *signed_tx)
 {
-	bitcoin_tx_input_set_witness(signed_tx, 0, NULL);
+	const char *err;
+	err = bitcoin_tx_input_set_witness(signed_tx, 0, NULL);
+	if (err)
+		fatal("%s: %s", __func__, err);
 }
 
```

**Comment by zerofeerouting** [comment]
Applied the patch, restarting CLN. 

Since the node already thinks it has published the FC transactions - how can I trigger the FC again?

**Comment by rustyrussell** [comment]
Hmm, that's another fix that went in v0.12,  but it's pretty easy to backport.  Here you are, it will just automatically re-generate and re-xmit on startup now:

```diff
diff --git a/lightningd/lightningd.c b/lightningd/lightningd.c
index b8e60b3ff..b689dd445 100644
--- a/lightningd/lightningd.c
+++ b/lightningd/lightningd.c
@@ -1150,6 +1150,12 @@ int main(int argc, char *argv[])
 			 "/dev/fd (if running in chroot) if you are "
 			 "approaching that many channels.");
 
+	/*~ If we have channels closing, make sure we re-xmit the last
+	 * transaction, in case bitcoind lost it. */
+	db_begin_transaction(ld->wallet->db);
+	resend_closing_transactions(ld);
+	db_commit_transaction(ld->wallet->db);
+
 	/*~ This is where we ask connectd to reconnect to any peers who have
 	 * live channels with us, and makes sure we're watching the funding
 	 * tx. */
diff --git a/lightningd/peer_control.c b/lightningd/peer_control.c
index bff93811c..caaaade29 100644
--- a/lightningd/peer_control.c
+++ b/lightningd/peer_control.c
@@ -292,6 +292,21 @@ void drop_to_chain(struct lightningd *ld, struct channel *channel,
 	resolve_close_command(ld, channel, cooperative);
 }
 
+void resend_closing_transactions(struct lightningd *ld)
+{
+	struct peer *peer;
+	struct channel *channel;
+
+	list_for_each(&ld->peers, peer, list) {
+		list_for_each(&peer->channels, channel, list) {
+			if (channel->state == CLOSINGD_COMPLETE)
+				drop_to_chain(ld, channel, true);
+			else if (channel->state == AWAITING_UNILATERAL)
+				drop_to_chain(ld, channel, false);
+		}
+	}
+}
+
 void channel_errmsg(struct channel *channel,
 		    struct peer_fd *peer_fd,
 		    const struct channel_id *channel_id UNUSED,
diff --git a/lightningd/peer_control.h b/lightningd/peer_control.h
index 77037e17d..77dd1a9cc 100644
--- a/lightningd/peer_control.h
+++ b/lightningd/peer_control.h
@@ -86,6 +86,9 @@ u8 *p2wpkh_for_keyidx(const tal_t *ctx, struct lightningd *ld, u64 keyidx);
 /* We've loaded peers from database, set them going. */
 void setup_peers(struct lightningd *ld);
 
+/* At startup, re-send any transactions we want bitcoind to have */
+void resend_closing_transactions(struct lightningd *ld);
+
 void drop_to_chain(struct lightningd *ld, struct channel *channel, bool cooperative);
 
 void channel_watch_funding(struct lightningd *ld, struct channel *channel);
diff --git a/lightningd/test/run-find_my_abspath.c b/lightningd/test/run-find_my_abspath.c
index a8805bce7..8ca265d0b 100644
--- a/lightningd/test/run-find_my_abspath.c
+++ b/lightningd/test/run-find_my_abspath.c
@@ -185,6 +185,9 @@ struct plugins *plugins_new(const tal_t *ctx UNNEEDED, struct log_book *log_book
 void plugins_set_builtin_plugins_dir(struct plugins *plugins UNNEEDED,
 				     const char *dir UNNEEDED)
 { fprintf(stderr, "plugins_set_builtin_plugins_dir called!\n"); abort(); }
+/* Generated stub for resend_closing_transactions */
+void resend_closing_transactions(struct lightningd *ld UNNEEDED)
+{ fprintf(stderr, "resend_closing_transactions called!\n"); abort(); }
 /* Generated stub for setup_color_and_alias */
 void setup_color_and_alias(struct lightningd *ld UNNEEDED)
 { fprintf(stderr, "setup_color_and_alias called!\n"); abort(); }
```

**Comment by zerofeerouting** [comment]
Thanks again @rustyrussell! Especially for this hint

> Hmm, that's another fix that went in v0.12

which made simply upgrade the node to the newest release. After upgrading the DB the node reissued all FCs, they're all `ONCHAIN` now.

Thanks again for the incredible support!

Since the problem is obviously solved by the upgrade, I'll close the issue.
